Firestore/Source/Core/FSTFirestoreClient.mm

```diff
@@ -111,8 +111,9 @@ - (instancetype)initWithDatabaseInfo:(const DatabaseInfo &)databaseInfo
           initialized = true;
           dispatch_semaphore_signal(initialUserAvailable);
         } else {
+          User userCopy = user;
```

Comment:
Okay, I've sat on this PR for a bit because I don't like this dance we're doing where the argument is a const reference and then we need to manually remember to make a copy and use it. I've seen several fix-up commits go by to catch all the places we didn't quite get this right.

It makes me extremely nervous.

I've reviewed all the callback implementations and I've concluded that we're probably much better off passing these by value instead of by const reference. For example, passing by value into this callback has the same number of copies as you've manually done and is far less error prone.

The assignment to the initialUser is a case where we possibly could have had an extra copy but we can offset by using move assignment.

My basic premise is that we're just not performance sensitive enough for the potential of an extra copy to matter and the safety we'll gain from doing so will be worth it.

For anything exceptionally heavyweight we need to either internalize a shared representation or just pass a std::shared_ptr. Passing references (or pointers) across callbacks is a recipe for disaster.

Reply:
Are you proposing this for just internal callbacks, or also the public api?

Reply:
For API elements that are truly callbacks (like listeners) I'm proposing by value there too, but we already know that the underling implementation of things like document snapshots will be a value type embedding a `std::shared_ptr`.

For futures, the future is returned by value but is reference counted under the covers. If the future allows returning its contents as `const T&` that's okay because the future owns the lifetime.

The problem we were getting into here was that we were passing references to temporaries despite the possibility of vastly different lifetimes.

The general rule I'd advocate is that if the callback escapes the caller's stack frame the callback should be defined to pass by value.

This means:
  * for-each style callbacks can be const reference because the lifetimes are expected to be fully contained by the caller of the for-each function.
  * pretty much everything else: by value